File: 2020/rs/src/bin/day_02.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("../input/input_02.txt").unwrap();
    let mut total_1 = 0u32;
    let mut total_2 = 0u32;
    for line in input.split('\n').filter(|s| s.len() > 0) {
        let tokens: Vec<_> = line.split(&[' ', '-'][..]).collect();
        let a: usize = tokens[0].parse().unwrap();
        let b: usize = tokens[1].parse().unwrap();
        let c: u8 = tokens[2].as_bytes()[0];
        let password = tokens[3].as_bytes();
        let c_count = password.iter().filter(|&&byte| byte == c).count();
        if a <= c_count && c_count <= b {
            total_1 += 1;
        }
        if (password[a-1] == c) != (password[b-1] == c) {
            total_2 += 1;
        }
    }
    println!("Part 1: {}", total_1);
    println!("Part 2: {}", total_2);
}

File: python/2020/day_02.py
import re
from pathlib import Path


def valid_1(a, b, char, password) -> bool:
    return a <= password.count(char) <= b


def valid_2(a, b, char, password) -> bool:
    return (password[a-1] == char) != (password[b-1] == char)


def main():
    text = (Path(__file__).parent / "../input/input_02.txt").read_text()
    total_1 = total_2 = 0
    for line in text.splitlines():
        a, b, char, password = re.split(r"[ :-]+", line)
        a, b = int(a), int(b)
        total_1 += valid_1(a, b, char, password)
        total_2 += valid_2(a, b, char, password)
    print("Part 1:", total_1)
    print("Part 2:", total_2)


if __name__ == "__main__":
    main()
